File: internal/model/progress.go
package model

import "time"

// ReadingProgress is the per-user, per-book read location.
// The authoritative position is (ChapterIdx, CharOffset) — see ADR-3.
// Percentage is a derived field, persisted only for quick UI display.
type ReadingProgress struct {
	ID         int       `json:"id"`
	UserID     int       `json:"-"`
	BookID     int       `json:"bookId"`
	ChapterIdx int       `json:"chapterIdx"`
	CharOffset int       `json:"charOffset"`
	Anchor     *string   `json:"anchor,omitempty"`
	ScrollPct  *float64  `json:"scrollPct,omitempty"`
	Percentage float64   `json:"percentage"`
	UpdatedAt  time.Time `json:"updatedAt"`
}

File: web/src/api/progress.js
import { api } from './client.js'

export const getProgress = (bookId) => api.get(`/books/${bookId}/progress`)

export const updateProgress = (bookId, { chapterIdx, charOffset, anchor, scrollPct, percentage }) =>
  api.put(`/books/${bookId}/progress`, { chapterIdx, charOffset, anchor, scrollPct, percentage })
